### ISSUE TITLE
error when executing asyncio.event.wait() via run_coroutine_threadsafe

### ISSUE THREAD

**Issue by rcarpa** [microsoft/pyright#5282]
Hi folks. I'm not yet sure that the following problem should be reported here or to `typeshed`. I ended reporting it here because the typeshed type hints "seem" correct to my untrained eye; and because mypy doesn't complain for the same code.

The following code:
```python
import asyncio

async def main():
    event = asyncio.Event()
    await asyncio.wrap_future(asyncio.run_coroutine_threadsafe(event.wait(), loop=asyncio.get_running_loop()))

asyncio.run(main())
```
generates an error with pyright 1.1.313 

```
error: Argument of type "Future[bool]" cannot be assigned to parameter "future" of type "concurrent.futures._base.Future[_T@wrap_future] | asyncio.futures.Future[_T@wrap_future]" in function "wrap_future"
    Type "Future[bool]" cannot be assigned to type "concurrent.futures._base.Future[_T@wrap_future] | asyncio.futures.Future[_T@wrap_future]"
      "Future[bool]" is incompatible with "Future[_T@wrap_future]"
        TypeVar "_T@Future" is invariant
          Type "Literal[True]" cannot be assigned to type "bool"
      "Future[bool]" is incompatible with "Future[_T@wrap_future]" (reportGeneralTypeIssues)
```

is this a problem on the pyright side?

**Comment by Hootner** [comment]
It seems that the issue might be with the Pyright type checker itself. Pyright might be inferring the type of asyncio.run_coroutine_threadsafe(event.wait(), loop=asyncio.get_running_loop()) as concurrent.futures.Future[bool] instead of asyncio.Future[bool]. This could lead to the type error you're encountering when you pass it to asyncio.wrap_future.

You can try explicitly annotating the type of the future to help Pyright understand the correct type:

python
Copy code
import asyncio
from asyncio import Future

async def main():
    event = asyncio.Event()
    future: Future[bool] = asyncio.run_coroutine_threadsafe(event.wait(), loop=asyncio.get_running_loop())
    await asyncio.wrap_future(future)

asyncio.run(main())
However, if you still encounter the issue with Pyright after trying this, it might be worth considering reporting the issue to the Pyright repository to get further assistance.It seems that the issue might be with the Pyright type checker itself. Pyright might be inferring the type of asyncio.run_coroutine_threadsafe(event.wait(), loop=asyncio.get_running_loop()) as concurrent.futures.Future[bool] instead of asyncio.Future[bool]. This could lead to the type error you're encountering when you pass it to asyncio.wrap_future.

You can try explicitly annotating the type of the future to help Pyright understand the correct type:

python
Copy code
import asyncio
from asyncio import Future

async def main():
    event = asyncio.Event()
    future: Future[bool] = asyncio.run_coroutine_threadsafe(event.wait(), loop=asyncio.get_running_loop())
    await asyncio.wrap_future(future)

asyncio.run(main())
However, if you still encounter the issue with Pyright after trying this, it might be worth considering reporting the issue to the Pyright repository to get further assistance.

**Comment by erictraut** [comment]
Here's a simplified repro that doesn't involve any complex overrides or async types:

```python
from typing import Awaitable, Literal, TypeVar

_T = TypeVar("_T")
class Future(Awaitable[_T]): ...
def func1(future: Future[_T]) -> Future[_T]: ...
def func2(coro: Awaitable[_T]) -> Future[_T]: ...
def func3() -> Awaitable[Literal[True]]: ...
func1(func2(func3()))
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
